src/Illuminate/Support/Siesta.php

```diff
@@ -0,0 +1,383 @@
+    {
+        $this->pending = $duration;
+
+        return $this;
+    }
+
+    /**
+     * Handle the object's destruction.
+     *
+     * @return void
+     */
+    public function __destruct()
+    {
+        if (! $this->shouldSleep) {
+            return;
+        }
+
+        if ($this->pending !== null) {
+            throw new RuntimeException('Unknown duration unit.');
+        }
+
+        if (static::$fake) {
+            static::$sequence[] = $this->duration;
+
+            return;
+        }
+
+        $remaining = $this->duration->copy();
+
+        $seconds = (int) $remaining->totalSeconds;
+
+        if ($seconds > 0) {
+            sleep($seconds);
+
+            $remaining = $remaining->subSeconds($seconds);
+        }
+
+        $microseconds = (int) $remaining->totalMicroseconds;
+
+        if ($microseconds > 0) {
+            usleep($microseconds);
+        }
+    }
+
+    /**
+     * Resolve the pending duration.
+     *
+     * @return int|float
+     */
+    protected function pullPending()
+    {
+        if ($this->pending === null) {
+            $this->shouldNotSleep();
+
+            throw new RuntimeException('No duration specified.');
+        }
+
+        if ($this->pending < 0) {
+            $this->pending = 0;
+        }
+
+        return tap($this->pending, function () {
+            $this->pending = null;
+        });
+    }
+
+    /**
+     * Stay awake and captured any attempts to sleep.
+     *
+     * @param  bool  $value
+     * @return void
+     */
+    public static function fake($value = true)
+    {
+        static::$fake = $value;
+
+        static::$sequence = [];
+    }
+
+    /**
+     * Assert the given sleeping occurred the a specific number of times.
+     *
+     * @param  \Closure  $expected
+     * @param  int  $times
+     * @return void
+     */
+    public static function assertSlept($expected, $times = 1)
+    {
+        $count = collect(static::$sequence)->filter($expected)->count();
+
+        PHPUnit::assertSame(
+            $times,
+            $count,
+            "The expected siesta was found [{$count}] times instead of [{$times}]."
+        );
+    }
+
+    /**
+     * Assert sleeping occurred the given times.
+     *
+     * @param  int  $expected
+     * @return void
+     */
+    public static function assertSleptTimes($expected)
+    {
+        PHPUnit::assertSame($expected, $count = count(static::$sequence), "Expected [{$expected}] siestas but found [{$count}].");
+    }
+
+    /**
+     * Assert the given sleep sequence was encountered.
+     *
+     * @param  array  $sequence
+     * @return void
+     */
+    public static function assertSequence($sequence)
+    {
+        static::assertSleptTimes(count($sequence));
+
+        collect($sequence)
+            ->zip(static::$sequence)
+            ->eachSpread(function (?Siesta $expected, CarbonInterval $actual) {
+                if ($expected === null) {
+                    return;
+                }
+
```

Comment:
There have been a code to prevent adding passed $sequence into static::$sequence in __destruct.
Is it ok for now to call assertion twice:
```
Seista::fake();
Seista::for(2)->seconds();
Siesta::assertSequence([
    Siesta::for(2)->seconds(), // won't it get into static::$sequence
]);
Siesta::assertSequence([
    Siesta::for(2)->seconds(),  // so this will fail
]);
```

Reply:
@Angel5a It is okay to call the assertion twice.

Any time you pass an instance through to an assertion method, we tell the internal class to not capture it.


```php
Siesta::fake();

Siesta::for(2)->seconds();

Siesta::assertSequence([
    Siesta::for(2)->seconds(),
]);
Siesta::assertSequence([
    Siesta::for(2)->seconds(),
]);

Siesta::assertSequence([
    Siesta::for(2)->seconds(),
    Siesta::for(3)->seconds(), << failure
]);
```

This test will fail where marked. The failure will indicate that you expected 2 sleeps, but it only found 1 - which is correct - as the others were for testing only.